File: pipeline/validate.py
"""JSON schema validation engine.

Loads each JSON data file, validates against its Pydantic model,
and reports errors with file path and context.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

from pydantic import TypeAdapter, ValidationError

from pipeline import DATA_DIR
from pipeline.models import FILE_MODELS


@dataclass
class ValidationResult:
    """Result of validating a single file."""

    file: str
    ok: bool
    record_count: int = 0
    errors: list[str] = field(default_factory=list)


def _preprocess_ships(data: dict[str, Any]) -> dict[str, Any]:
    """Rename 'class' → 'hullClass' in ship hulls (reserved word workaround)."""
    if "hulls" in data:
        for hull in data["hulls"]:
            if "class" in hull:
                hull["hullClass"] = hull.pop("class")
    return data


def validate_file(filename: str, data_dir: Path = DATA_DIR) -> ValidationResult:
    """Validate a single JSON file against its Pydantic model."""
    filepath = data_dir / filename

    if filename not in FILE_MODELS:
        return ValidationResult(file=filename, ok=False, errors=[f"No model registered for {filename}"])

    if not filepath.exists():
        return ValidationResult(file=filename, ok=False, errors=[f"File not found: {filepath}"])

    model_cls, is_bare_array = FILE_MODELS[filename]

    try:
        raw = json.loads(filepath.read_text(encoding="utf-8"))
    except json.JSONDecodeError as e:
        return ValidationResult(file=filename, ok=False, errors=[f"Invalid JSON: {e}"])

    # Pre-processing for known quirks
    if filename == "ships.json" and isinstance(raw, dict):
        raw = _preprocess_ships(raw)

    try:
        if is_bare_array:
            adapter = TypeAdapter(list[model_cls])
            result = adapter.validate_python(raw)
            count = len(result)
        else:
            result = model_cls.model_validate(raw)
            # Try to count records from the main list field
            count = _count_records(result)

        return ValidationResult(file=filename, ok=True, record_count=count)

    except ValidationError as e:
        errors = []
        for err in e.errors():
            loc = " → ".join(str(x) for x in err["loc"])
            errors.append(f"  {loc}: {err['msg']}")
        return ValidationResult(file=filename, ok=False, errors=errors)


def _count_records(model: object) -> int:
    """Attempt to count the primary records in a validated model."""
    # Try common field names that contain the main data list
    for attr in ("races", "classes", "feats", "skills", "exaltations", "backgrounds", "alignments", "packages", "hulls", "weapons"):
        val = getattr(model, attr, None)
        if isinstance(val, list):
            return len(val)
        if isinstance(val, dict):
            # For skills.json which has nested dicts of lists
            total = sum(len(v) for v in val.values() if isinstance(v, list))
            if total > 0:
                return total
    return 0


def validate_all(data_dir: Path = DATA_DIR) -> list[ValidationResult]:
    """Validate all registered JSON files."""
    results = []
    for filename in sorted(FILE_MODELS.keys()):
        results.append(validate_file(filename, data_dir))
    return results


def cross_reference_check(data_dir: Path = DATA_DIR) -> list[str]:
    """Check cross-file references for consistency.

    Verifies:
    - Class skill names exist in skills.json
    - Class feat names exist in feats.json
    - NPC trait names exist in traits.json
    """
    issues: list[str] = []

    try:
        skills_raw = json.loads((data_dir / "skills.json").read_text(encoding="utf-8"))
        skill_names = set()
        for group_list in skills_raw.get("skills", {}).values():
            for skill in group_list:
                skill_names.add(skill["name"])

        classes_raw = json.loads((data_dir / "classes.json").read_text(encoding="utf-8"))
        for cls in classes_raw.get("classes", []):
            for skill_name in cls.get("skills", []):
                if skill_name not in skill_names:
                    issues.append(f"classes.json: class '{cls['name']}' references unknown skill '{skill_name}'")
    except Exception as e:
        issues.append(f"Cross-ref check (classes→skills) failed: {e}")

    try:
        feats_raw = json.loads((data_dir / "feats.json").read_text(encoding="utf-8"))
        feat_names = {f["name"] for f in feats_raw.get("feats", [])}

        classes_raw = json.loads((data_dir / "classes.json").read_text(encoding="utf-8"))
        for cls in classes_raw.get("classes", []):
            for feat_entry in cls.get("feats", []):
                feat_name = feat_entry["name"] if isinstance(feat_entry, dict) else feat_entry

                # Handle choice-based entries like "Two Weapon Fighting OR Far Shot"
                if " OR " in feat_name:
                    alternatives = [alt.strip() for alt in feat_name.split(" OR ")]
                    found_any = False
                    for alt in alternatives:
                        base = alt.split(" (")[0] if " (" in alt else alt
                        if base in feat_names or alt in feat_names:
                            found_any = True
                            break
                    if not found_any:
                        issues.append(f"classes.json: class '{cls['name']}' references unknown feat '{feat_name}'")
                    continue

                # Feats with parenthetical variants like "Skill Focus (Any Lore)" won't match exactly
                # Strip parenthetical for base name matching
                base_name = feat_name.split(" (")[0] if " (" in feat_name else feat_name
                if base_name not in feat_names and feat_name not in feat_names:
                    issues.append(f"classes.json: class '{cls['name']}' references unknown feat '{feat_name}'")
    except Exception as e:
        issues.append(f"Cross-ref check (classes→feats) failed: {e}")

    try:
        traits_raw = json.loads((data_dir / "traits.json").read_text(encoding="utf-8"))
        trait_names = {t["name"] for t in traits_raw}
        trait_ids = {t["id"] for t in traits_raw}

        npcs_raw = json.loads((data_dir / "npc-templates.json").read_text(encoding="utf-8"))
        for npc in npcs_raw:
            for trait_ref in npc.get("traits", []):
                # Trait refs can be dicts {"id": "fear", "param": 3} or strings
                if isinstance(trait_ref, dict):
                    trait_id = trait_ref.get("id", "")
                    if trait_id not in trait_ids:
                        issues.append(f"npc-templates.json: NPC '{npc['name']}' references unknown trait id '{trait_id}'")
                else:
                    # String refs may have parameters like "Fear (3)" — strip those
                    base_trait = trait_ref.split(" (")[0] if " (" in trait_ref else trait_ref
                    if base_trait not in trait_names:
                        issues.append(f"npc-templates.json: NPC '{npc['name']}' references unknown trait '{trait_ref}'")
    except Exception as e:
        issues.append(f"Cross-ref check (npcs→traits) failed: {e}")

    return issues

File: scripts/validate.ts
/**
 * TypeScript port of pipeline/validate.py + pipeline/cli.py (validate command).
 *
 * Validates all 12 JSON data files against Zod schemas and optionally
 * runs cross-reference checks.
 *
 * Usage:
 *   bun run scripts/validate.ts          # validate all
 *   bun run scripts/validate.ts --xref   # also cross-reference check
 */

import * as fs from "node:fs";
import * as path from "node:path";
import type { ZodError } from "zod";
import { FILE_SCHEMAS, type SchemaEntry } from "../src/lib/dtd/schemas/index.ts";

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

const SCRIPT_DIR = path.dirname(decodeURIComponent(new URL(import.meta.url).pathname).replace(/^\/([A-Z]:)/i, "$1"));
const DATA_DIR = path.resolve(SCRIPT_DIR, "..", "data");

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface ValidationResult {
	file: string;
	ok: boolean;
	recordCount: number;
	errors: string[];
}

// ---------------------------------------------------------------------------
// Record counting
// ---------------------------------------------------------------------------

/** Known top-level array fields that contain the "main" records. */
const LIST_FIELDS = [
	"races",
	"classes",
	"feats",
	"skills",
	"exaltations",
	"backgrounds",
	"alignments",
	"packages",
	"hulls",
	"weapons",
] as const;

export function countRecords(data: unknown): number {
	if (Array.isArray(data)) return data.length;
	if (typeof data !== "object" || data === null) return 0;

	const obj = data as Record<string, unknown>;
	for (const field of LIST_FIELDS) {
		const val = obj[field];
		if (Array.isArray(val)) return val.length;
		// skills.json has nested dicts of arrays
		if (typeof val === "object" && val !== null && !Array.isArray(val)) {
			let total = 0;
			for (const v of Object.values(val)) {
				if (Array.isArray(v)) total += v.length;
			}
			if (total > 0) return total;
		}
	}
	return 0;
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

export function validateFile(filename: string): ValidationResult {
	const entry: SchemaEntry | undefined = FILE_SCHEMAS[filename];
	if (!entry) {
		return { file: filename, ok: false, recordCount: 0, errors: [`No schema registered for ${filename}`] };
	}

	const filepath = path.join(DATA_DIR, filename);
	if (!fs.existsSync(filepath)) {
		return { file: filename, ok: false, recordCount: 0, errors: [`File not found: ${filepath}`] };
	}

	let raw: unknown;
	try {
		raw = JSON.parse(fs.readFileSync(filepath, "utf-8"));
	} catch (e) {
		return { file: filename, ok: false, recordCount: 0, errors: [`Invalid JSON: ${e}`] };
	}

	const result = entry.schema.safeParse(raw);
	if (result.success) {
		return { file: filename, ok: true, recordCount: countRecords(result.data), errors: [] };
	}

	const zodErr = result.error as ZodError;
	const errors = zodErr.issues.map((issue) => {
		const loc = issue.path.join(".");
		return `    ${loc} — ${issue.message} (${issue.code})`;
	});
	return { file: filename, ok: false, recordCount: 0, errors };
}

export function validateAll(): ValidationResult[] {
	return Object.keys(FILE_SCHEMAS)
		.sort()
		.map((filename) => validateFile(filename));
}

// ---------------------------------------------------------------------------
// Cross-reference checks
// ---------------------------------------------------------------------------

export function loadJson(filename: string): unknown {
	return JSON.parse(fs.readFileSync(path.join(DATA_DIR, filename), "utf-8"));
}

export function crossReferenceCheck(): string[] {
	const issues: string[] = [];

	// 1. classes → skills
	try {
		const skillsRaw = loadJson("skills.json") as Record<string, unknown>;
		const skillNames = new Set<string>();
		const skillsMap = (skillsRaw.skills ?? {}) as Record<string, Array<{ name: string }>>;
		for (const groupList of Object.values(skillsMap)) {
			for (const skill of groupList) {
				skillNames.add(skill.name);
			}
		}

		const classesRaw = loadJson("classes.json") as Record<string, unknown>;
		const classes = (classesRaw.classes ?? []) as Array<{ name: string; skills: string[] }>;
		for (const cls of classes) {
			for (const skillName of cls.skills ?? []) {
				if (!skillNames.has(skillName)) {
					issues.push(`classes.json: class '${cls.name}' references unknown skill '${skillName}'`);
				}
			}
		}
	} catch (e) {
		issues.push(`Cross-ref check (classes→skills) failed: ${e}`);
	}

	// 2. classes → feats
	try {
		const featsRaw = loadJson("feats.json") as Record<string, unknown>;
		const featsArr = (featsRaw.feats ?? []) as Array<{ name: string }>;
		const featNames = new Set(featsArr.map((f) => f.name));

		const classesRaw = loadJson("classes.json") as Record<string, unknown>;
		const classes = (classesRaw.classes ?? []) as Array<{
			name: string;
			feats: Array<string | { name: string }>;
		}>;

		for (const cls of classes) {
			for (const featEntry of cls.feats ?? []) {
				const featName = typeof featEntry === "string" ? featEntry : featEntry.name;

				// Handle "Two Weapon Fighting OR Far Shot" OR-alternatives
				if (featName.includes(" OR ")) {
					const alternatives = featName.split(" OR ").map((s) => s.trim());
					const foundAny = alternatives.some((alt) => {
						const base = alt.includes(" (") ? alt.split(" (")[0] : alt;
						return featNames.has(base) || featNames.has(alt);
					});
					if (!foundAny) {
						issues.push(`classes.json: class '${cls.name}' references unknown feat '${featName}'`);
					}
					continue;
				}

				// Strip parenthetical variants like "Skill Focus (Any Lore)"
				const baseName = featName.includes(" (") ? featName.split(" (")[0] : featName;
				if (!featNames.has(baseName) && !featNames.has(featName)) {
					issues.push(`classes.json: class '${cls.name}' references unknown feat '${featName}'`);
				}
			}
		}
	} catch (e) {
		issues.push(`Cross-ref check (classes→feats) failed: ${e}`);
	}

	// 3. npcs → traits
	try {
		const traitsRaw = loadJson("traits.json") as Array<{ id: string; name: string }>;
		const traitNames = new Set(traitsRaw.map((t) => t.name));
		const traitIds = new Set(traitsRaw.map((t) => t.id));

		const npcsRaw = loadJson("npc-templates.json") as Array<{
			name: string;
			traits: Array<string | { id: string; param?: unknown }>;
		}>;

		for (const npc of npcsRaw) {
			for (const traitRef of npc.traits ?? []) {
				if (typeof traitRef === "object" && traitRef !== null) {
					const traitId = (traitRef as { id: string }).id ?? "";
					if (!traitIds.has(traitId)) {
						issues.push(`npc-templates.json: NPC '${npc.name}' references unknown trait id '${traitId}'`);
					}
				} else {
					// String refs may have parameters like "Fear (3)" — strip those
					const baseTrait =
						typeof traitRef === "string" && traitRef.includes(" (")
							? traitRef.split(" (")[0]
							: String(traitRef);
					if (!traitNames.has(baseTrait)) {
						issues.push(`npc-templates.json: NPC '${npc.name}' references unknown trait '${traitRef}'`);
					}
				}
			}
		}
	} catch (e) {
		issues.push(`Cross-ref check (npcs→traits) failed: ${e}`);
	}

	return issues;
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

function main(): void {
	const args = process.argv.slice(2);
	const xref = args.includes("--xref");

	// ── Schema validation ──────────────────────────────────────────────
	const results = validateAll();

	console.log("\nSchema Validation");
	console.log("─".repeat(50));

	let allOk = true;
	for (const r of results) {
		const padded = r.file.padEnd(24);
		if (r.ok) {
			console.log(`  ${padded} ✓   ${r.recordCount} records`);
		} else {
			allOk = false;
			console.log(`  ${padded} ✗`);
			for (const err of r.errors.slice(0, 5)) {
				console.log(err);
			}
			if (r.errors.length > 5) {
				console.log(`    ... and ${r.errors.length - 5} more`);
			}
		}
	}

	// ── Cross-reference checks ─────────────────────────────────────────
	if (xref) {
		console.log("\nCross-reference checks:");
		const issues = crossReferenceCheck();
		if (issues.length > 0) {
			for (const issue of issues) {
				console.log(`  ⚠ ${issue}`);
			}
			allOk = false;
		} else {
			console.log("  ✓ All cross-references valid");
		}
	}

	// ── Summary ────────────────────────────────────────────────────────
	const passed = results.filter((r) => r.ok).length;
	const total = results.length;

	if (allOk) {
		console.log(`\nAll ${total} files validated successfully.\n`);
	} else {
		console.log(`\n${total - passed}/${total} files failed validation.\n`);
		process.exit(1);
	}
}

// Only run when executed directly (not imported by tests)
const isDirectRun = process.argv[1] && /validate\.ts$/.test(process.argv[1]);
if (isDirectRun) {
	main();
}
